Fix constant folding bug in jnp.arange for non-zero start values

yashwantbezawada opened the pull request:
This fixes issue #32542 where `jnp.arange` was incorrectly folding into a constant array when called with a non-zero start value.

**Problem:**
When calling `jnp.arange(1, n)`, JAX was creating a constant array instead of using the `iota` primitive. This caused compilation time to explode for large arrays because the entire array was embedded in the computation graph.

**Root cause:**
The optimization path in `_arange` only applied `lax.broadcasted_iota` when `start == 0`. For any other start value, it fell back to `array(np.arange(...))`, which creates a concrete numpy array first and then converts it to a JAX array, triggering constant folding.

**Fix:**
Extended the optimization to handle any start value:
- Use `lax.broadcasted_iota` to generate the base sequence [0, 1, 2, ...]
- When start != 0, add the offset using `lax.add`
- This ensures the arange is computed using primitives rather than folded into constants

**Testing:**
The fix correctly handles:
- Zero start values (existing behavior preserved)
- Non-zero start values (bug fix)
- Empty ranges (size <= 0)
- Float start/stop values (using ceil)

Fixes #32542

jakevdp reviewed (changes requested):
Thanks for this! It looks reasonable, but we will need test coverage for the new logic.

In particular, you could add a sample to `testArangeJaxpr` to confirm that the new code produces a single call to `iota_p` with nonzero startpoint.

Also, we should add another test case covering some of the potential corner cases with the new implementation, such as the correctness of `np.ceil` in your implementation.

jakevdp commented on tests/lax_numpy_test.py:
```diff
@@ -4855,20 +4855,62 @@ def testArangeJit(self):
-    num_eqs = 2 if device is not None else 1
+
+    # Determine expected number of equations
+    start = 0 if len(args) == 1 else args[0]
```
Tests should have minimal logic. Rather than putting this logic in the test, we should write a separate test for the case where the start value is non-zero.

jakevdp commented on tests/lax_numpy_test.py:
```diff
@@ -4855,20 +4855,62 @@ def testArangeJit(self):
+    """Test corner cases with np.ceil in arange implementation."""
+    test_cases = [
+      # Float values where ceiling affects result
+      (1.5, 5.5),   # ceil(4.0) = 4
```
The comments showing what `ceil` is in each case are probably not necessary. I think we could achieve more-or-less the same test coverage much more concisely by generating random start and stop values using `jtu.rand_default`, for `float32` and `int32` dtypes.

yashwantbezawada replied:
Good point - I've replaced this with randomized testing for more concise coverage.

jakevdp reviewed (commented):
Thanks! A couple more changes requested below.

yashwantbezawada commented on tests/lax_numpy_test.py:
```diff
@@ -4855,20 +4855,62 @@ def testArangeJit(self):
+        jax_result = jnp.arange(start, stop)
+        np_result = np.arange(start, stop)
+        self.assertArraysEqual(jax_result, np_result)
+
```
Will do - I've added test coverage for complex arguments.

jakevdp reviewed (commented):
Looks good, thanks!

Last thing: can you please squash your changes into a single commit? See https://docs.jax.dev/en/latest/contributing.html#single-change-commits-and-pull-requests

yashwantbezawada committed:
Fix constant folding bug in jnp.arange for non-zero start

When jnp.arange is called with a non-zero start value (e.g.,
jnp.arange(1, n)), the array was being folded into a constant instead
of using the iota primitive. This caused compilation time to explode
for large arrays.

Extended the optimization to handle any start value by computing the
array size as ceil(stop - start) and using iota with an offset added
when start is non-zero.

Also adds a check to skip the optimization when start or stop is
complex, since ceil doesn't support complex numbers.

Test coverage includes:
- testArangeJaxprNonZeroStart: Verifies non-zero start uses iota + add
- testArangeRandomValues: Parameterized randomized testing for
  int32/float32 dtypes
- testArangeComplex: Ensures complex arguments match NumPy behavior

Fixes #32542

yashwantbezawada commented:
Done! Squashed all commits into a single commit as requested.

jakevdp reviewed (approved):
Thanks!

jakevdp reviewed (approved):
Thanks!
